app/services/users/create.rb

```diff
@@ -7,7 +7,19 @@ def initialize(user_params:, shop_params:)
     end
 
     def call
-      # TODO: write your code here. You can inspire by this code: https://github.com/mateuszbialowas/brug-service_objects/pull/1/files
+      fail! :shop_invalid if CreateShopValidator.new.call(@shop_params).failure?
+      fail! :user_invalid if CreateUserValidator.new.call(@user_params).failure?
+
+      ActiveRecord::Base.transaction do
+        @shop = Shop.create!(@shop_params)
+        @user = User.create!(@user_params.merge(shop: @shop))
+      end
+
+      RegistrationMailer.with(user: @user).welcome_email.deliver_later
+    rescue => e
+      fail! e.to_sym
+    ensure
+      @result = {shop: @shop, user: @user}
```

Comment:
Chyba brakuje mi 
```
def add_result(**hash)
    result.merge!(**hash)
  end
```

zeby result móc napychać co krok - co wydaje mi sie pozwolic moze na lepsze debugowanie i wydaje mi sie wygodniejsze niz agregowanie tego na koniec kodu. Chociaz tutaj jestem otwarty na dyskusje, bo coś czuje, że coś mi umyka ;]

Reply:
Ja mam odwrotnie. Wynik na dobrą sprawę jest tylko jeden. Ale to też dlatego, że ja mam do tego trochę inne podejście. Bo mam wrażenie, że wy kombinujecie w kierunku: "jest wynik albo błąd". Jak dla mnie to serwis może zakończyć się błędem i wciąż zawracać wyniki, np: napchane błędami walidacyjnymi modele albo szczegóły błędu.

Reply:
Jesli zwraca błąd to fajnie widziec w ktorym momencie i co sie po drodze wydarzylo. Dlatego podzial an result/errors jest imho spoko. I jesli napychany result po drodze to mozna miec te informacje latwo dostepne. Ale to ze to na papierze dobrze brzmi nie znaczy ze faktycznie byloby uzyteczne ;]